src/librustc_metadata/decoder.rs

```diff
@@ -831,7 +831,16 @@ impl<'a, 'tcx> CrateMetadata {
                             let ctor_def_id = self.get_ctor_def_id(child_index).unwrap_or(def_id);
                             let ctor_kind = self.get_ctor_kind(child_index);
                             let ctor_def = Def::Ctor(ctor_def_id, CtorOf::Variant, ctor_kind);
-                            let vis = self.get_visibility(ctor_def_id.index);
+                            let mut vis = self.get_visibility(ctor_def_id.index);
+                            // If the variant is marked as non_exhaustive then lower the visibility
+                            // to within the crate.
```

Comment:
Could you add to this comment that this logic exists for fictive constructors (where `unwrap_or` applies above) as the visibility of non-exhaustive const or fn ctor will have been encoded correctly?

Alternatively, instead of calling `unwrap_or` on the `ctor_def_id`, we could match on it and apply this logic in the `None` arm which might better indicate the cases that this logic is intended to address.